src/async_traits.rs

```diff
@@ -48,40 +64,175 @@ where
         Func: FnOnce(&mut Conn) -> Result<R, E> + Send + 'static,
     {
-        let connection = self.get_owned_connection().await;
-        Self::run_with_connection(connection, f).await
+        let connection = self.get_owned_connection();
+        connection.run_with_connection(f).await
     }
 
     #[doc(hidden)]
-    async fn run_with_connection<R, E, Func>(
-        connection: Self::OwnedConnection,
-        f: Func,
-    ) -> Result<R, E>
+    async fn run_with_connection<R, E, Func>(self, f: Func) -> Result<R, E>
     where
         R: Send + 'static,
         E: Send + 'static,
         Func: FnOnce(&mut Conn) -> Result<R, E> + Send + 'static,
     {
-        spawn_blocking(move || f(&mut *Self::as_sync_conn(&connection)))
+        spawn_blocking(move || f(&mut *self.as_sync_conn()))
             .await
             .unwrap() // Propagate panics
     }
 
     #[doc(hidden)]
-    async fn run_with_shared_connection<R, E, Func>(
-        connection: Arc<Self::OwnedConnection>,
-        f: Func,
-    ) -> Result<R, E>
+    async fn run_with_shared_connection<R, E, Func>(self: &Arc<Self>, f: Func) -> Result<R, E>
     where
         R: Send + 'static,
         E: Send + 'static,
         Func: FnOnce(&mut Conn) -> Result<R, E> + Send + 'static,
     {
-        spawn_blocking(move || f(&mut *Self::as_sync_conn(&connection)))
+        let conn = self.clone();
+        spawn_blocking(move || f(&mut *conn.as_sync_conn()))
             .await
             .unwrap() // Propagate panics
     }
 
+    #[doc(hidden)]
+    async fn transaction_depth(&self) -> Result<u32, DieselError> {
+        let conn = self.get_owned_connection();
+
+        Self::run_with_connection(conn, |conn| {
+            match Conn::TransactionManager::transaction_manager_status_mut(&mut *conn) {
+                TransactionManagerStatus::Valid(status) => {
+                    return Ok(status.transaction_depth().map(|d| d.into()).unwrap_or(0));
+                }
+                TransactionManagerStatus::InError => {
+                    return Err(DieselError::BrokenTransactionManager);
+                }
+            }
+        })
+        .await
+    }
+
+    #[doc(hidden)]
+    async fn start_transaction(self: &Arc<Self>) -> Result<(), DieselError> {
+        if self.transaction_depth().await? != 0 {
+            return Err(DieselError::AlreadyInTransaction);
+        }
+        self.run_with_shared_connection(|conn| Conn::TransactionManager::begin_transaction(conn))
+            .await?;
+        Ok(())
+    }
+
+    #[doc(hidden)]
+    async fn add_retry_savepoint(self: &Arc<Self>) -> Result<(), DieselError> {
+        match self.transaction_depth().await? {
+            0 => return Err(DieselError::NotInTransaction),
+            1 => (),
+            _ => return Err(DieselError::AlreadyInTransaction),
+        };
+
+        self.run_with_shared_connection(|conn| Conn::TransactionManager::begin_transaction(conn))
+            .await?;
+        Ok(())
+    }
```

Comment:
So admittedly "BEGIN" or "SAVEPOINT ..." are both invoked through `begin_transaction`, but I figured I'd make the separation explicit, with some validation of the current transaction depth.

(I also find this makes the body of `transaction_async_with_retry` more readable)

Reply:
This could use a comment - without knowing this, it looks like `begin_transaction` is being used twice instead of something like `savepoint`. Is the behaviour like

```
if in a transaction at depth 1 {
  SAVEPOINT
} else if not in a transaction {
  BEGIN
} else {
  // error
}
```

?

Reply:
It's documented as a part of Diesel's API: https://docs.diesel.rs/master/diesel/connection/trait.TransactionManager.html#tymethod.begin_transaction

> Begin a new transaction or savepoint
>
> If the transaction depth is greater than 0, this should create a savepoint instead. This function is expected to increment the transaction depth by 1.

Diesel would be happy to keep adding savepoints, but CRDB isn't: that's why I validate the transaction depth manually + split this into two different helpers for BEGIN / SAVEPOINT.

- Diesel is using savepoints called `diesel_savepoint_<DEPTH>` (See: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#342 )
- In contrast, CRDB wants the SAVEPOINT to be called `cockroach_restart`
- However, CRDB provides a tool to accept any savepoint name (including the one chosen by Diesel): https://www.cockroachlabs.com/docs/v23.1/advanced-client-side-transaction-retries#customizing-the-retry-savepoint-name
- **BUT** this disallows nested transactions -- which is why I'm checking the depth here. Technically I could omit my depth checks, but IMO this results in a more confusing error, because CRDB throws errors at runtime.

Either way I can add some comments expanding on what's happening here.

Reply:
Comments added